docs/design/sparse-v2.md

```diff
@@ -0,0 +1,225 @@
+
+In particular, the path rules for sparse patterns for _always_ be repo paths,
+not client paths. Thus, if the client wants to track "foo" and rename it to
+"subdir/bar", they must `jj sparse set --add foo` and
+`jj client-map set --from foo --to bar`. In other words, the mapping operation can
+be thought of as always _after_ the sparse operation.
+
+New commands will enable editing of the WorkingCopyMap:
+
+- `jj client-map list` will print all mapping pairs.
+- `jj client-map set --from foo --to bar` will add a new mapping to the end of the list.
+- `jj client-map remove --from foo` will remove a specific mapping rule.
+- `jj client-map edit` will pull up a text editor for manual editing.
+
+Like sparse paths, mapping rules are defined to apply in _order_ and on any
+save operation will be modified to a minimal canonical form. Thus,
+`jj client-map set --from "" --to ""` will always completely wipe the map.
+The first matching rule in reverse list order determines how a particular
+folder/file should be named in the client.
+
+For simplicity, we require that all client-maps are _invertible_, that is:
+
+- For any repo path R, there is at most one associated client path C
+- Having no associated client path is fine (that part of the repo will never be materialized)
+- For any client path C, there is at most one associated repo path R
+- Having no associated repo path is fine (all files in such paths will be ignored)
+
+To this end, recursive mappings are considers _destructive_. That is, if "foo"
+is renamed to "bar", but "bar" already exists in the repo, then the mapping rule
+hides the original "bar" directory, making it uneditable in the client.
+
+## Versioning and Storage
+
+Updating the active Sparse Pattern for a particular working copy will now
+take place in two separate steps, one which updates the Op Store, and
+another which updates the actual working copy.
+
+This gives the user the ability to update the active sparse patterns whilst
+not interacting with the local working copy, which is useful for custom
+integrations which may not be _able_ to check out particular sparse patterns due
+to problems with the backend (encoding, permission errors, etc.). A bad
+`jj sparse set --add oops` command can thus be undone, even via `jj op undo`
+if desired.
+
+### View Updates
+
+The View object will be migrated to store sparse patterns via id. The
+indirection will save on storage since sparse patterns are not expected to
+change very frequently.
+
+```rust
+// Before:
+pub wc_commit_ids: HashMap<WorkspaceId, CommitId>,
+
+// After:
+pub struct WorkingCopyInfo {
+    pub commit_id: CommitId,
+    pub sparse_patterns_id: SparsePatternsId,
+}
+...
+pub wc_info: HashMap<WorkspaceId, WorkingCopyInfo>,
+```
+
+A repo with no sparse patterns in View storage will be actively migrated upon
+running the first jj command with the new release, filling the curring active
+SparsePatternsId into all historical Views. This will be maintained for ~4
```

Comment:
I was thinking you wouldn't be able to checkout old commits this way, but that's a separate layer from the View object... yea, this seems like it would be fine (fill in the current patterns when loading a View that doesn't have them).